spec/controller/admin/tours_controller_spec.rb

```diff
@@ -0,0 +1,232 @@
+require "rails_helper"
+include SessionsHelper
+
+RSpec.shared_examples "update examples" do
+  it "should found" do
+    expect(response).to  have_http_status 302
+  end
+
+  it "should show success flash" do
+    expect(flash[:success]).to  eql I18n.t(".admin.tours.update.passed")
+  end
+end
+
+RSpec.describe Admin::ToursController, type: :controller do
+  let(:tour){ FactoryBot.create :tour }
+  let(:user){ FactoryBot.create :user }
+
+  before do
+    log_in user
+  end
+
+  describe "GET #show" do
+    context "when found tour" do
+      before do
+        get :show, params: {id: tour.id}
+        sleep 1
+      end
+
+      it "assigns @tour" do
+        expect(assigns(:tour)).to eql (tour)
+      end
+
+      it "return a success response" do
+        expect(response).to have_http_status(200)
+      end
+
+      it "render show template" do
+        expect(response).to render_template "admin/tours/show"
+      end
+    end
+
+    context "when not found tour" do
+      before do
+        get :show, params: {id: 2}
+        sleep 1
+      end
+
+      it "return a found status" do
+        expect(response).to have_http_status(302)
+      end
+
+      it "redirect to admin root path" do
+        expect(response).to redirect_to admin_root_path
+      end
+
+      it "show flash danger" do
+        expect(flash[:danger]).to eql I18n.t("admin.tours.show.not_found")
+      end
+    end
+  end
+
+  describe "GET #new"do
+    context "when get new tour" do
+      before do
+        get :new
+      end
+      it "return a OK status" do
+        expect(response).to have_http_status(200)
+      end
+
+      it "render a new template" do
+        expect(response).to render_template "admin/tours/new"
+      end
+    end
+  end
+
+  describe "POST #create" do
+    context "when successful create" do
+      before do
+        allow_any_instance_of(Tour).to receive(:validate_image).and_return(true)
```

Comment:
phân biệt let vs let!
phân biệt mock và stub

Reply:
phân biệt let với let!
- khi ta gọi biến nhiều lần trong một example thì nó chỉ khai báo khối let một lần duy nhất thôi vào lần gọi đầu tiên.
- let! sẽ được gọi trong một khối before ngầm, do đó kết quả được tính toán và cached lại trước khi khối it thực thi.
phân biệt mock và stub
- mock thể hiện những sự mong đợi khi gọi đến một method nào đó, nó có thể là kết quả trả về, là đối số truyền vào method đó hay là số lần gọi đến method đó.
- stub là định nghĩa/cho phép một object gọi đến một method và trả về một kết quả nào đó, khi example của ta không gọi đến method đó thì cũng không sao cả.

Reply:
dùng mock để mong đợi số lần gọi thì ntn

Reply:
nếu dùng mock để mong đợi số lần gọi thì khi chúng ta viết một example thì số lần gọi của đối tượng đó phải đúng như ta mong đợi, nếu không thì nó sẽ trả về kết quả test thất bại